File: lib/Models/BookingModels.dart
import 'BaseResponse.dart';
import 'FoodModels.dart';
import 'BranchModels.dart';
import 'PromotionModels.dart';

// ==================== BookingTableModel ====================
class BookingTableModel {
  BookingTableModel({
    required this.id,
    required this.tableCode,
    required this.capacity,
    this.location,
    this.status,
  });

  final String id;
  final String tableCode;
  final int capacity;
  final String? location;
  final String? status;

  factory BookingTableModel.fromJson(Map<String, dynamic> json) {
    return BookingTableModel(
      id: json['id'] as String,
      tableCode: json['tableCode'] as String,
      capacity: (json['capacity'] as num).toInt(),
      location: json['location'] as String?,
      status: json['status'] as String?,
    );
  }
}

// ==================== TableAvailabilityModel ====================
class TableAvailabilityModel {
  TableAvailabilityModel({
    required this.tableId,
    required this.tableCode,
    required this.capacity,
    required this.status,
    required this.booked,
    this.reservedFrom,
    this.reservedTo,
  });

  final String tableId;
  final String tableCode;
  final int capacity;
  final String status;
  final bool booked;
  final DateTime? reservedFrom;
  final DateTime? reservedTo;

  factory TableAvailabilityModel.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(String? input) =>
        input == null ? null : DateTime.parse(input);

    return TableAvailabilityModel(
      tableId: json['tableId'] as String,
      tableCode: json['tableCode'] as String,
      capacity: (json['capacity'] as num).toInt(),
      status: json['status'] as String,
      booked: json['booked'] as bool,
      reservedFrom: parseDate(json['reservedFrom'] as String?),
      reservedTo: parseDate(json['reservedTo'] as String?),
    );
  }
}

// ==================== BookingDishModel ====================
class BookingDishModel {
  BookingDishModel({
    required this.id,
    required this.quantity,
    required this.unitPrice,
    required this.servingOrder,
    this.specialNote,
    this.food,
  });

  final String id;
  final int quantity;
  final int unitPrice;
  final int servingOrder;
  final String? specialNote;
  final FoodModel? food;

  int get totalPrice => quantity * unitPrice;

  factory BookingDishModel.fromJson(Map<String, dynamic> json) {
    final food = json['food'] as Map<String, dynamic>?;
    return BookingDishModel(
      id: json['id'] as String,
      quantity: (json['quantity'] as num).toInt(),
      unitPrice: (json['unitPrice'] as num).toInt(),
      servingOrder: (json['servingOrder'] as num).toInt(),
      specialNote: json['specialNote'] as String?,
      food: food != null ? FoodModel.fromJson(food) : null,
    );
  }
}

// ==================== BookingDishPayload ====================
class BookingDishPayload {
  BookingDishPayload({
    required this.foodId,
    required this.quantity,
    required this.servingOrder,
    this.specialNote,
  });

  final String foodId;
  final int quantity;
  final int servingOrder;
  final String? specialNote;

  Map<String, dynamic> toJson() => {
        'foodId': foodId,
        'quantity': quantity,
        'servingOrder': servingOrder,
        if (specialNote != null && specialNote!.isNotEmpty)
          'specialNote': specialNote,
      };
}

// ==================== BookingRequestPayload ====================
class BookingRequestPayload {
  BookingRequestPayload({
    required this.bookingTime,
    required this.guests,
    required this.tableId,
    required this.branchId,
    this.specialRequest,
    this.dishes = const [],
    this.promotionCode,
    this.durationMinutes = 120,
  });

  final DateTime bookingTime;
  final int guests;
  final String tableId;
  final String branchId;
  final String? specialRequest;
  final List<BookingDishPayload> dishes;
  final String? promotionCode;
  final int durationMinutes;

  Map<String, dynamic> toJson() => {
        'bookingTime': bookingTime.toIso8601String(),
        'guests': guests,
        'tableId': tableId,
        'branchId': branchId,
        'durationMinutes': durationMinutes,
        if (specialRequest != null && specialRequest!.isNotEmpty)
          'specialRequest': specialRequest,
        if (dishes.isNotEmpty) 'dishes': dishes.map((e) => e.toJson()).toList(),
        if (promotionCode != null && promotionCode!.isNotEmpty)
          'promotionCode': promotionCode,
      };
}

// ==================== BookingResponseModel ====================
class BookingResponseModel {
  BookingResponseModel({
    required this.id,
    required this.reservedFrom,
    required this.reservedTo,
    required this.guests,
    required this.status,
    this.specialRequest,
    this.branch,
    this.table,
    this.subtotalAmount,
    this.discountAmount,
    this.totalAmount,
    this.promotion,
    this.dishes = const [],
  });

  final String id;
  final DateTime reservedFrom;
  final DateTime reservedTo;
  final int guests;
  final String status;
  final String? specialRequest;
  final BranchModel? branch;
  final BookingTableModel? table;
  final int? subtotalAmount;
  final int? discountAmount;
  final int? totalAmount;
  final PromotionModel? promotion;
  final List<BookingDishModel> dishes;

  String get branchName => branch?.name ?? '';
  String get tableCode => table?.tableCode ?? '';

  int get computedSubtotal {
    if (subtotalAmount != null) return subtotalAmount!;
    return dishes.fold<int>(0, (sum, dish) => sum + dish.totalPrice);
  }

  int get computedDiscount => discountAmount ?? 0;

  int get computedTotal {
    if (totalAmount != null) return totalAmount!;
    final value = computedSubtotal - computedDiscount;
    return value < 0 ? 0 : value;
  }

  factory BookingResponseModel.fromJson(Map<String, dynamic> json) {
    final table = json['table'] as Map<String, dynamic>?;
    final branch = json['branch'] as Map<String, dynamic>?;
    final dishes = json['dishes'] as List<dynamic>? ?? [];
    final promotion = json['promotion'] as Map<String, dynamic>?;
    return BookingResponseModel(
      id: json['id'] as String,
      reservedFrom: DateTime.parse(json['reservedFrom'] as String),
      reservedTo: DateTime.parse(json['reservedTo'] as String),
      guests: (json['guests'] as num).toInt(),
      status: json['status'] as String,
      specialRequest: json['specialRequest'] as String?,
      branch: branch != null ? BranchModel.fromJson(branch) : null,
      table: table != null ? BookingTableModel.fromJson(table) : null,
      subtotalAmount: (json['subtotalAmount'] as num?)?.toInt(),
      discountAmount: (json['discountAmount'] as num?)?.toInt(),
      totalAmount: (json['totalAmount'] as num?)?.toInt(),
      promotion: promotion != null ? PromotionModel.fromJson(promotion) : null,
      dishes: dishes
          .map((e) => BookingDishModel.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

// ==================== BookingListResponse ====================
class BookingListResponse extends BaseResponse {
  late List<BookingResponseModel> data;

  BookingListResponse({
    required int statusCode,
    required String message,
    List<BookingResponseModel> data = const [],
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory BookingListResponse.fromJson(Map<String, dynamic> json) {
    final raw = json['data'];
    final list = raw is List ? raw : const [];
    return BookingListResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(BookingResponseModel.fromJson)
          .toList(),
    );
  }
}

// ==================== BookingResponseWrapper ====================
class BookingResponseWrapper extends BaseResponse {
  BookingResponseModel? data;

  BookingResponseWrapper({
    required int statusCode,
    required String message,
    this.data,
  }) : super(statusCode: statusCode, message: message);

  factory BookingResponseWrapper.fromJson(Map<String, dynamic> json) {
    final payload = json['data'];
    return BookingResponseWrapper(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: payload is Map<String, dynamic>
          ? BookingResponseModel.fromJson(payload)
          : null,
    );
  }
}

// ==================== TableAvailabilityListResponse ====================
class TableAvailabilityListResponse extends BaseResponse {
  late List<TableAvailabilityModel> data;

  TableAvailabilityListResponse({
    required int statusCode,
    required String message,
    List<TableAvailabilityModel> data = const [],
  }) : super(statusCode: statusCode, message: message) {
    this.data = data;
  }

  factory TableAvailabilityListResponse.fromJson(Map<String, dynamic> json) {
    final raw = json['data'];
    final list = raw is List ? raw : const [];
    return TableAvailabilityListResponse(
      statusCode: (json['statusCode'] as num).toInt(),
      message: json['message'] as String,
      data: list
          .whereType<Map<String, dynamic>>()
          .map(TableAvailabilityModel.fromJson)
          .toList(),
    );
  }
}

File: mobile_restaurant/models/BookingModels.ts
import { parseBaseResponse, type BaseResponse } from './AuthModels';
import { parseFoodModel, type FoodModel } from './FoodModels';
import { parseBranchModel, type BranchModel } from './BranchModels';
import { parsePromotionModel, type PromotionModel } from './PromotionModels';

// ==================== BookingTableModel ====================
export interface BookingTableModel {
  id: string;
  tableCode: string;
  capacity: number;
  location?: string;
  status?: string;
}

export function parseBookingTableModel(json: any): BookingTableModel {
  return {
    id: json.id,
    tableCode: json.tableCode,
    capacity: Number(json.capacity),
    location: json.location ?? undefined,
    status: json.status ?? undefined,
  };
}

// ==================== TableAvailabilityModel ====================
export interface TableAvailabilityModel {
  tableId: string;
  tableCode: string;
  capacity: number;
  status: string;
  booked: boolean;
  reservedFrom?: string;
  reservedTo?: string;
}

export function parseTableAvailabilityModel(json: any): TableAvailabilityModel {
  return {
    tableId: json.tableId,
    tableCode: json.tableCode,
    capacity: Number(json.capacity),
    status: json.status,
    booked: json.booked,
    reservedFrom: json.reservedFrom ?? undefined,
    reservedTo: json.reservedTo ?? undefined,
  };
}

// ==================== BookingDishModel ====================
export interface BookingDishModel {
  id: string;
  quantity: number;
  unitPrice: number;
  servingOrder: number;
  specialNote?: string;
  food?: FoodModel;
  totalPrice: number;
}

export function parseBookingDishModel(json: any): BookingDishModel {
  const qty = Number(json.quantity);
  const price = Number(json.unitPrice);
  return {
    id: json.id,
    quantity: qty,
    unitPrice: price,
    servingOrder: Number(json.servingOrder),
    specialNote: json.specialNote ?? undefined,
    food: json.food ? parseFoodModel(json.food) : undefined,
    totalPrice: qty * price,
  };
}

// ==================== BookingDishPayload ====================
export interface BookingDishPayload {
  foodId: string;
  quantity: number;
  servingOrder: number;
  specialNote?: string;
}

// ==================== BookingRequestPayload ====================
export interface BookingRequestPayload {
  bookingTime: string; // ISO 8601
  guests: number;
  tableId: string;
  branchId: string;
  specialRequest?: string;
  dishes?: BookingDishPayload[];
  promotionCode?: string;
  durationMinutes?: number;
}

// ==================== BookingResponseModel ====================
export interface BookingResponseModel {
  id: string;
  reservedFrom: string;
  reservedTo: string;
  guests: number;
  status: string;
  specialRequest?: string;
  branch?: BranchModel;
  table?: BookingTableModel;
  subtotalAmount?: number;
  discountAmount?: number;
  totalAmount?: number;
  promotion?: PromotionModel;
  dishes: BookingDishModel[];
  branchName: string;
  tableCode: string;
  computedSubtotal: number;
  computedDiscount: number;
  computedTotal: number;
  rated: boolean;
  depositAmount?: number;
  depositRefunded?: boolean;
}

export function parseBookingResponseModel(json: any): BookingResponseModel {
  const dishes = (json.dishes as any[] ?? []).map(parseBookingDishModel);
  const subtotalAmount = json.subtotalAmount != null ? Number(json.subtotalAmount) : undefined;
  const discountAmount = json.discountAmount != null ? Number(json.discountAmount) : undefined;
  const totalAmount = json.totalAmount != null ? Number(json.totalAmount) : undefined;
  const depositAmount = json.depositAmount != null ? Number(json.depositAmount) : undefined;
  const depositRefunded = json.depositRefunded !== undefined ? !!json.depositRefunded : undefined;

  const computedSubtotal = subtotalAmount ?? dishes.reduce((sum, d) => sum + d.totalPrice, 0);
  const computedDiscount = discountAmount ?? 0;
  const computedTotal = totalAmount ?? Math.max(0, computedSubtotal - computedDiscount);

  return {
    id: json.id,
    reservedFrom: json.reservedFrom,
    reservedTo: json.reservedTo,
    guests: Number(json.guests),
    status: json.status,
    specialRequest: json.specialRequest ?? undefined,
    branch: json.branch ? parseBranchModel(json.branch) : undefined,
    table: json.table ? parseBookingTableModel(json.table) : undefined,
    subtotalAmount,
    discountAmount,
    totalAmount,
    depositAmount,
    depositRefunded,
    promotion: json.promotion ? parsePromotionModel(json.promotion) : undefined,
    dishes,
    branchName: json.branch?.name ?? '',
    tableCode: json.table?.tableCode ?? '',
    computedSubtotal,
    computedDiscount,
    computedTotal,
    rated: !!json.rated,
  };
}

// ==================== BookingListResponse ====================
export interface BookingListResponse extends BaseResponse {
  data: BookingResponseModel[];
}

export function parseBookingListResponse(json: any): BookingListResponse {
  const base = parseBaseResponse(json);
  const raw = json.data;
  const list = Array.isArray(raw) ? raw : [];
  return {
    ...base,
    data: list.map(parseBookingResponseModel),
  };
}

// ==================== BookingResponseWrapper ====================
export interface BookingResponseWrapper extends BaseResponse {
  data?: BookingResponseModel;
}

export function parseBookingResponseWrapper(json: any): BookingResponseWrapper {
  const base = parseBaseResponse(json);
  return {
    ...base,
    data: json.data && typeof json.data === 'object' ? parseBookingResponseModel(json.data) : undefined,
  };
}

// ==================== TableAvailabilityListResponse ====================
export interface TableAvailabilityListResponse extends BaseResponse {
  data: TableAvailabilityModel[];
}

export function parseTableAvailabilityListResponse(json: any): TableAvailabilityListResponse {
  const base = parseBaseResponse(json);
  const raw = json.data;
  const list = Array.isArray(raw) ? raw : [];
  return {
    ...base,
    data: list.map(parseTableAvailabilityModel),
  };
}
